python/packaging/frontend_sdist/setup.py

```diff
@@ -15,42 +15,96 @@
 # limitations under the License.
 #
 
+import os
+import subprocess
 import sys
 
 from setuptools import setup
 from setuptools.command.install import install
-import subprocess as sp
 
 tensorrt_module = "##TENSORRT_MODULE##"
+tensorrt_version = "##TENSORRT_PYTHON_VERSION##"
+tensorrt_submodules = [
+    "{}_libs=={}".format(tensorrt_module, tensorrt_version),
+    "{}_bindings=={}".format(tensorrt_module, tensorrt_version),
+]
 
 
 class InstallCommand(install):
     def run(self):
-        def install_dep(package_name):
-            status = sp.run(
-                [
-                    sys.executable,
-                    "-m",
-                    "pip",
-                    "install",
-                    "{:}==##TENSORRT_PYTHON_VERSION##".format(package_name),
-                    "--index-url",
-                    "https://pypi.nvidia.com",
-                ]
-            )
-            status.check_returncode()
-
-        install_dep("{:}_libs".format(tensorrt_module))
-        install_dep("{:}_bindings".format(tensorrt_module))
-
-        install.run(self)
+        # pip-inside-pip hack ref #3080
+        subprocess.check_call(
+            [
+                "{}/bin/pip".format(sys.exec_prefix),
+                "install",
+                "--extra-index-url",
+                "https://pypi.nvidia.com",
+                *tensorrt_submodules,
+            ]
+        )
+
+        super().run()
+
+
+def pip_config_list():
+    """Get the current pip config (env vars, config file, etc)."""
+    return subprocess.check_output(
+        [
+            "{}/bin/pip".format(sys.exec_prefix),
+            "config",
+            "list",
+        ]
+    ).decode()
+
+
+def parent_command_line():
+    """Get the command line of the parent PID."""
+    pid = os.getppid()
+    # try retrieval using psutil
+    try:
+        import psutil
+
+        return " ".join(psutil.Process(pid).cmdline())
+    except ModuleNotFoundError:
+        pass
+    # fall back to shell
+    try:
+        return (
+            subprocess.check_output(["ps", "-p", str(pid), "-o", "command"])
+            .decode()
+            .split("\n")[1]
+        )
+    except subprocess.CalledProcessError:
+        return ""
+
+
+# use pip-inside-pip hack only if the nvidia index is not set in the environment
+if "pypi.nvidia.com" in pip_config_list() or "pypi.nvidia.com" in parent_command_line():
```

Comment:
A recent internal (and unrelated) discussion revealed we may want one additional condition here. In order to test different versions or from staging servers (other than `pypi.nvidia.com`) a full override may be desired. Basically a "manage the source and installation of dependencies yourself" option. I think one way to handle this would be to check an environment variable when set assumes you have set `--extra-index-url` to something you know works, which could be something other than `pypi.nvidia.com`. This may be useful externally as well if someone is running their own internal PyPI server and mirroring our wheel/sdist files as-is. @pranavm-nvidia what do you think about this approach and calling this environment variable something like `NVIDIA_TENSORRT_DISABLE_INTERNAL_PIP`?

Reply:
how about we make `NVIDIA_INDEX_URL` configurable? defaulting to `https://pypi.nvidia.com`, but it could be overwritten with custom index urls or set it to an empty string to always pass the test?

Reply:
0f13ecd

Reply:
I think `NVIDIA_TENSORRT_DISABLE_INTERNAL_PIP` is still useful even if we make the NVIDIA pip index URL configurable. There could be a case where there are no extra indexes used (not sure how it would work, but I imagine it would be possible somehow).

Reply:
2c382eb